Simulator/simulator.py

```diff
@@ -95,324 +85,687 @@ def __init__(self, etype: etype, time: datetime, job_id: int):
+        while self.current_reserve_jobs:
+            finish_time, _, job_id, node_ids = self.current_reserve_jobs[0]
+            if finish_time <= current_time:
+                for node_id in node_ids:
+                    # Remove the job from the node
+                    self._reserve_nodes[node_id].remove_job(job_id)
+                # Pop the job from the running job list
+                self.current_reserve_jobs.pop(0)
             else:
-              break
-      next_event = self._get_next_event()
-
-    t_end = time.time()
-    print("Time taken in seconds is ", t_end - t_start)
-
-    # Post- Simulation: Add schedule column and wait_time column to input dataframe
-    self.input_trace["schedule"] = (
-      self.input_trace["job_id"].map(self.job_schedule_map).fillna("D")
-    )
-    self.input_trace["wait_time_sec"] = (
-      self.input_trace["job_id"].map(self.wait_time_map).fillna(0)
-    )
-
-    # Steady state
-    h = 0.1  # head
-    steady_state = self.input_trace[
-      (self.input_trace.job_id >= h * self.job_count)
-      & (self.input_trace.job_id <= (1.0 - h) * self.job_count)
-      ]
-    mean_wait_time_sec = steady_state[
-      steady_state.schedule == "R"
-      ].wait_time_sec.mean()
-    balking_rate = (
-            1
-            - self.input_trace[self.input_trace.schedule == "R"].time.count()
-            / self.input_trace.time.count()
-    )
-
-    # LJW and Compound policy -- On-demand jobs don't wait in queue
-    mean_wait_time_sec = mean_wait_time_sec * (1 - balking_rate)
-
-    # Simulation time
-    elapsed_simulation_time = (end - start).total_seconds()
-
-    # User has not specified trace duration
-    if duration_hrs is None:
-      duration_hrs = elapsed_simulation_time / 3600
-
-    # Computing total cost
-    (
-      total_cost,
-      reserved_cost,
-      on_demand_cost,
-      effective_price,
-    ) = self._compute_total_cost(self.input_trace, duration_hrs)
-
-    print(
-      "Reserved cost is ",
-      reserved_cost,
-      ", On demand cost is ",
-      on_demand_cost,
-      ", Total cost is ",
-      reserved_cost + on_demand_cost,
-      )
-    print("----------------------------------------------------------------")
-
-    return (mean_wait_time_sec, effective_price, balking_rate, total_cost)
+                break
+
+
+    def resubmit_long_jobs_after_spec_execution(self, job, runtime_threshold_sec):
+        # Resubmit job after running for "t" time per speculative execution
+        job.loc['speculative_execution'] = True
+        job.loc['submit_time'] = job['submit_time'] + runtime_threshold_sec
+
+        # Place job in correct place in trace
+        temp_df = self.input_trace[self.input_trace['submit_time'] > job.submit_time]
+        new_index = temp_df.iloc[0].job_id
+        job.loc['job_id'] = new_index
+
+        job = job.to_frame().T
+
+        # Reorder and re-index input trace to reflect new timing of job
```

Comment:
Nice! This makes sense conceptually. I'm not a python expert, but looking at the comments, this seems like the right move

Reply:
thanks! yeah, took a bit of experimentation in IDLE to make sure these operations were right, but it seems to be working properly